### PR TITLE
Replace use of openssl_certificate with a custom certgen Ansible role.

### PR DESCRIPTION
**Pull request by dgoodwin**
Replace openssl_certificate with custom certgen role.
    
    The cert module in Ansible is a no-go on RHEL, and we don't want to rule
    that out yet. Replace our older apiserver certgen script with a role we
    can use for both apiserver certs, and individual cluster certs.
    
    Generates a CA if one does not exist and uses it for all certs.
    
    Each cluster gets it's own cert to prevent problems with the same cert
    being uploaded anywhere in AWS.

### CODE REVIEW

File: contrib/ansible/create-cluster-playbook.yaml [hunk 1]
```diff
@@ -37,9 +37,11 @@
 
     # Paths where we will generate self-signed certs for the specific cluster. (probably don't need to
     # override these but if you wanted to use pre-existing certs you could)
-    cluster_cert_path: "{{ playbook_dir }}/../../certs/{{ cluster_name }}.pem"
-    cluster_privatekey_path: "{{ playbook_dir }}/../../certs/{{ cluster_name }}-key.pem"
-    cluster_csr_path: "{{ playbook_dir }}/../../certs/{{ cluster_name }}.csr"
+    cluster_cert_dir: "{{ playbook_dir }}/../../certs"
```
Comment (csrwng) [inline contrib/ansible/create-cluster-playbook.yaml]: add this dir to .gitignore

File: contrib/ansible/create-cluster-playbook.yaml [hunk 2]
```diff
@@ -82,25 +84,22 @@
...
-      provider: selfsigned
+  - include_role:
+      name: cluster_operator_certgen
+    vars:
+      cert_dir: "{{ cluster_cert_dir }}"
+      ca_name: "{{ cluster_ca_name }}"
+      cert_name: "{{ cluster_name }}"
+      # TODO: do something meaningful here?
+      cn: "{{ cluster_name }}"
+      alt_names: "{{ cluster_name }}.{{ cluster_namespace }}\",\"{{ cluster_name }}.{{ cluster_namespace }}.svc"
```
Comment (csrwng) [inline contrib/ansible/create-cluster-playbook.yaml]: How does this work? If I expand this, I would get a string like:
`cewong-cluster.myproject","cewong-cluster.myproject.svc`

File: contrib/ansible/deploy-playbook.yaml [hunk 3]
```diff
@@ -44,23 +50,30 @@
...
-      creates: "{{ playbook_dir }}/../../apiserver.pem"
-    when: redeploy_apiserver_cert | bool
+  - include_role:
+      name: cluster_operator_certgen
+    vars:
+      cert_dir: "{{ apiserver_cert_dir }}"
+      ca_name: "{{ apiserver_ca_name }}"
+      cert_name: "apiserver"
+      cn: cluster-operator-apiserver
+      alt_names: "cluster-operator-apiserver.{{ cluster_operator_namespace }}\",\"cluster-operator-apiserver.{{ cluster_operator_namespace }}.svc"
```
Comment (csrwng) [inline contrib/ansible/deploy-playbook.yaml]: same comment as above

File: contrib/ansible/roles/cluster_operator_certgen/templates/cert-config.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"CN":"{{ cn }}","hosts":["{{ alt_names }}"],"key":{"algo":"rsa","size":2048}}
```
Comment (csrwng) [inline contrib/ansible/roles/cluster_operator_certgen/templates/cert-config.json]: use .j2 suffix

File: contrib/ansible/roles/cluster_operator_certgen/templates/ca-config.json [hunk 5]
```diff
@@ -0,0 +1 @@
+{"signing":{"default":{"expiry":"43800h","usages":["signing","key encipherment","server"]}}}
```
Comment (csrwng) [inline contrib/ansible/roles/cluster_operator_certgen/templates/ca-config.json]: If not a template, it should likely go in role/[name]/files

File: contrib/ansible/roles/cluster_operator_certgen/templates/ca-setup.json [hunk 6]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (csrwng) [inline contrib/ansible/roles/cluster_operator_certgen/templates/ca-setup.json]: use .j2 suffix

File: contrib/ansible/roles/cluster_operator_certgen/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    dest: "{{ cert_config_file }}"
+  when: not cert_stat.stat.exists
+
+- name: generate cert
+  shell: "cfssl gencert -ca={{ ca_file }} -ca-key={{ ca_key_file }} -config={{ ca_config_file }} {{ cert_config_file }} | cfssljson -bare {{ cert_name }}"
+  args:
+    chdir: "{{ cert_dir }}"
+  when: not cert_stat.stat.exists
+
+#- name: cleanup temporary files
```
Comment (csrwng) [inline contrib/ansible/roles/cluster_operator_certgen/tasks/main.yml]: Is there a reason you commented out this part?

File: contrib/examples/create-cluster.sh [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#! /bin/bash
```
Comment (csrwng) [inline contrib/examples/create-cluster.sh]: delete this file

File: contrib/ansible/create-cluster-playbook.yaml (hunk 2 above)
Comment (dgoodwin) [reply contrib/ansible/create-cluster-playbook.yaml]: Pretty sure it's all just junk today, this is just modeled on how it was working before, and we were actually using the apiserver's cert in any case. Definitely needs to be looked at so added the TODO for both this and the CN.

File: contrib/ansible/deploy-playbook.yaml (hunk 3 above)
Comment (dgoodwin) [reply contrib/ansible/deploy-playbook.yaml]: Not sure just following what was in the apiserver cert gen script.

File: contrib/ansible/roles/cluster_operator_certgen/tasks/main.yml (hunk 7 above)
Comment (dgoodwin) [reply contrib/ansible/roles/cluster_operator_certgen/tasks/main.yml]: Just debugging so I could diff these with the old variants, actually exposed the problem. Readding the cleanup next commit.

File: .gitignore [hunk 9]
```diff
@@ -22,12 +22,5 @@ contrib/examples/*-cluster.yaml
...
 .var
 .vscode
-apiserver-key.pem
-apiserver.csr
-apiserver.pem
-cluster-operator-ca.json
-ca-key.pem
-ca.csr
-ca.pem
-server-ca-config.json
```
Comment (dgoodwin) [inline .gitignore]: Devs can clean these up, they are no longer used.

**Comment by dgoodwin** [comment]
Updated for feedback and cleaned up.

**Pull request merged**